File: packages/store/src/fuzzy.ts
/**
 * Approximate term matching: over the corpus index ({@link pickVariants}) and
 * over small hand-written texts such as notes ({@link matchText}).
 *
 * # What problem this solves
 *
 * The corpus is OCR output. A name the source spells correctly on every page
 * reaches the index as `sindicato` 269 times and as `sindigato`, `sinicato` or
 * `bindicato` a handful of times each. An exact search misses exactly the
 * pages a historian is least likely to find any other way.
 *
 * # Why the vocabulary, and why frequency decides
 *
 * Variants are drawn from the index's own vocabulary (`fts_items_vocab`), so
 * nothing is suggested that no document contains, and proper names absent
 * from any dictionary are covered for free.
 *
 * Edit distance alone is not enough: `pescado` is one edit from `pesado`, and
 * both are real words. What separates a misreading from a different word is
 * weight. Measured on a real corpus, an OCR error is far rarer than the word
 * it corrupts, while a different word — or a plural — is of comparable
 * weight. So a variant is kept only when it is clearly rarer than the term
 * (a misreading of it), or clearly more common (the term is itself the
 * misspelling, typed by the writer or read by the OCR). Everything of similar
 * weight is left alone.
 */

/** A variant must differ in document frequency by at least this factor. */
const WEIGHT_RATIO = 4

/** Misreadings one term may add. Past this the query drowns in noise. */
const MAX_MISREADINGS = 8

/**
 * The same folding the `unicode61 remove_diacritics 1` tokenizer applies, so a
 * query term can be looked up in the vocabulary it produced.
 */
export function normalizeTerm(term: string): string {
  return term.normalize('NFD').replace(/\p{M}/gu, '').toLowerCase()
}

/**
 * How many edits a term tolerates. Short terms none — one edit turns `mar`
 * into `mas` — and nothing with a digit, because dates, folios and signatures
 * are only useful exact.
 */
export function maxEditsFor(term: string): number {
  if (/\d/.test(term)) return 0
  if (term.length < 5) return 0
  if (term.length < 8) return 1
  return 2
}

/**
 * Optimal-string-alignment distance: substitutions, insertions, deletions and
 * adjacent transpositions each cost one. Returns `max + 1` as soon as the
 * distance is known to exceed `max`, which is what keeps a scan over the
 * whole vocabulary cheap.
 */
export function editDistance(a: string, b: string, max: number): number {
  if (Math.abs(a.length - b.length) > max) return max + 1

  let before: number[] | null = null
  let previous = Array.from({ length: b.length + 1 }, (_, j) => j)

  for (let i = 1; i <= a.length; i++) {
    const current = [i]
    let rowMin = i
    for (let j = 1; j <= b.length; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1
      let value = Math.min(previous[j]! + 1, current[j - 1]! + 1, previous[j - 1]! + cost)
      if (before && i > 1 && j > 1 && a[i - 1] === b[j - 2] && a[i - 2] === b[j - 1]) {
        value = Math.min(value, before[j - 2]! + 1)
      }
      current.push(value)
      if (value < rowMin) rowMin = value
    }
    if (rowMin > max) return max + 1
    before = previous
    previous = current
  }

  return previous[b.length]!
}

interface Candidate {
  term: string
  distance: number
  docs: number
}

function byCloseness(x: Candidate, y: Candidate): number {
  return x.distance - y.distance || y.docs - x.docs || x.term.localeCompare(y.term)
}

/**
 * The vocabulary terms worth searching alongside `term`.
 *
 * `vocab` maps each indexed term to the number of documents holding it.
 */
export function pickVariants(term: string, vocab: ReadonlyMap<string, number>): string[] {
  const needle = normalizeTerm(term)
  const max = maxEditsFor(needle)
  if (max === 0) return []

  const weight = vocab.get(needle) ?? 0
  const misreadings: Candidate[] = []
  const corrections: Candidate[] = []

  for (const [candidate, docs] of vocab) {
    if (candidate === needle || Math.abs(candidate.length - needle.length) > max) continue
    const distance = editDistance(needle, candidate, max)
    if (distance > max) continue

    if (docs * WEIGHT_RATIO <= weight) {
      misreadings.push({ term: candidate, distance, docs })
    } else if (docs >= Math.max(2, weight * WEIGHT_RATIO)) {
      corrections.push({ term: candidate, distance, docs })
    }
  }

  // Only the best-supported correction: a typo has one intended word, and a
  // second candidate is usually a different word that happens to be close.
  const correction = corrections.sort(byCloseness).slice(0, 1)
  const kept = misreadings.sort(byCloseness).slice(0, MAX_MISREADINGS)
  return [...correction, ...kept].map((candidate) => candidate.term)
}

/** How a text answered a query: as written, only through typos, or not at all. */
export type TextMatch = 'exact' | 'approximate' | null

/**
 * Whether `text` holds every word of `query`, in any order, ignoring case and
 * accents.
 *
 * A word matches as written when the text contains it, even inside a longer
 * word — searching `molin` still finds `molineros`, as the literal search this
 * replaces always did. Failing that, it matches with a typo on either side
 * when some word of the text is within the edits {@link maxEditsFor} allows.
 *
 * For small, hand-written texts such as notes, where every candidate can be
 * read. It is not an index: over the corpus, use the FTS vocabulary instead.
 */
export function matchText(query: string, text: string): TextMatch {
  const terms = normalizeTerm(query).split(/\s+/).filter(Boolean)
  if (terms.length === 0) return null

  const haystack = normalizeTerm(text)
  let words: string[] | null = null
  let approximate = false

  for (const term of terms) {
    if (haystack.includes(term)) continue
    const max = maxEditsFor(term)
    if (max === 0) return null
    words ??= haystack.split(/[^\p{L}\p{N}]+/u).filter(Boolean)
    if (!words.some((word) => editDistance(term, word, max) <= max)) return null
    approximate = true
  }

  return approximate ? 'approximate' : 'exact'
}

File: apps/desktop/src-tauri/src/nlp/fuzzy.rs
//! Approximate term matching for the lexical legs of retrieval.
//!
//! The Rust twin of `packages/store/src/fuzzy.ts`, which holds the reasoning
//! in full. In short: variants come from the index's own vocabulary
//! (`fts_items_vocab`, migration 0037) and are kept only when their document
//! frequency marks them as a misreading of the term (far rarer), the spelling a
//! typo meant (far more common), or — when both are rare — another spelling of
//! a rare name. The two implementations must agree; a test below reads the
//! TypeScript constants and fails if they drift.

use std::collections::HashMap;

use rusqlite::Connection;

/// A variant must differ in document frequency by at least this factor.
const WEIGHT_RATIO: i64 = 4;
/// Misreadings one term may add.
const MAX_MISREADINGS: usize = 8;
/// Below this many documents a count says nothing about spelling.
const RARE_DOCS: i64 = 10;

/// The app setting that turns approximate search off ('off'). Shared with the
/// TypeScript side (`SETTINGS_KEYS.SEARCH_FUZZY`).
pub const SEARCH_FUZZY_SETTING: &str = "search_fuzzy";

const VOCAB_SQL: &str = "SELECT term, doc FROM fts_items_vocab \
     WHERE length(term) >= 4 AND term NOT GLOB '*[0-9]*'";

/// Folds case and the diacritics of Latin letters, as the index tokenizer
/// (`unicode61 remove_diacritics 1`) does. A table rather than a Unicode
/// normalization crate: the corpus is Spanish, and this covers every letter it
/// writes without adding a dependency.
pub fn normalize_term(term: &str) -> String {
    term.chars()
        .flat_map(char::to_lowercase)
        .map(|c| match c {
            'á' | 'à' | 'â' | 'ä' | 'ã' | 'å' => 'a',
            'é' | 'è' | 'ê' | 'ë' => 'e',
            'í' | 'ì' | 'î' | 'ï' => 'i',
            'ó' | 'ò' | 'ô' | 'ö' | 'õ' => 'o',
            'ú' | 'ù' | 'û' | 'ü' => 'u',
            'ñ' => 'n',
            'ç' => 'c',
            'ý' | 'ÿ' => 'y',
            other => other,
        })
        .collect()
}

/// Edits a term tolerates: none under 5 letters or with a digit, one up to 7
/// letters, two from 8.
pub fn max_edits_for(term: &str) -> usize {
    let length = term.chars().count();
    if term.chars().any(|c| c.is_ascii_digit()) || length < 5 {
        0
    } else if length < 8 {
        1
    } else {
        2
    }
}

/// Optimal-string-alignment distance, giving up at `max + 1`.
pub fn edit_distance(a: &str, b: &str, max: usize) -> usize {
    let a: Vec<char> = a.chars().collect();
    let b: Vec<char> = b.chars().collect();
    if a.len().abs_diff(b.len()) > max {
        return max + 1;
    }
    let mut before: Option<Vec<usize>> = None;
    let mut previous: Vec<usize> = (0..=b.len()).collect();
    for i in 1..=a.len() {
        let mut current = vec![i; b.len() + 1];
        let mut row_min = i;
        for j in 1..=b.len() {
            let cost = usize::from(a[i - 1] != b[j - 1]);
            let mut value = (previous[j] + 1)
                .min(current[j - 1] + 1)
                .min(previous[j - 1] + cost);
            if let Some(before) = &before {
                if i > 1 && j > 1 && a[i - 1] == b[j - 2] && a[i - 2] == b[j - 1] {
                    value = value.min(before[j - 2] + 1);
                }
            }
            current[j] = value;
            row_min = row_min.min(value);
        }
        if row_min > max {
            return max + 1;
        }
        before = Some(previous);
        previous = current;
    }
    previous[b.len()]
}

struct Candidate<'a> {
    term: &'a str,
    distance: usize,
    docs: i64,
}

fn by_closeness(x: &Candidate, y: &Candidate) -> std::cmp::Ordering {
    x.distance
        .cmp(&y.distance)
        .then(y.docs.cmp(&x.docs))
        .then(x.term.cmp(y.term))
}

/// The vocabulary terms worth searching alongside `term`.
pub fn pick_variants(term: &str, vocab: &HashMap<String, i64>) -> Vec<String> {
    let needle = normalize_term(term);
    let max = max_edits_for(&needle);
    if max == 0 {
        return Vec::new();
    }
    let needle_len = needle.chars().count();
    let weight = vocab.get(&needle).copied().unwrap_or(0);
    let mut misreadings = Vec::new();
    let mut corrections = Vec::new();

    for (candidate, &docs) in vocab {
        if *candidate == needle || candidate.chars().count().abs_diff(needle_len) > max {
            continue;
        }
        let distance = edit_distance(&needle, candidate, max);
        if distance > max {
            continue;
        }
        let entry = Candidate {
            term: candidate,
            distance,
            docs,
        };
        if docs * WEIGHT_RATIO <= weight {
            misreadings.push(entry);
        } else if docs >= 2.max(weight * WEIGHT_RATIO) {
            corrections.push(entry);
        } else if weight > 0 && weight < RARE_DOCS && docs < RARE_DOCS {
            misreadings.push(entry);
        }
    }

    corrections.sort_by(by_closeness);
    misreadings.sort_by(by_closeness);
    corrections
        .into_iter()
        .take(1)
        .chain(misreadings.into_iter().take(MAX_MISREADINGS))
        .map(|candidate| candidate.term.to_string())
        .collect()
}

/// Whether approximate search is on. Anything but 'off' — including a setting
/// that cannot be read — counts as on.
pub fn fuzzy_enabled(conn: &Connection) -> bool {
    crate::settings::get_setting(conn, SEARCH_FUZZY_SETTING).as_deref() != Some("off")
}

/// An FTS5 expression matching any close variant of `terms` and none of the
/// terms themselves, or `None` when approximate search is off, the vocabulary
/// is unavailable, or no term has a variant.
///
/// Variants only, on purpose: the caller has already run the exact search,
/// and a variant — rare by nature — would outrank the real word under BM25 if
/// both went into one query. Asked separately, its finds can be placed after.
pub fn variants_only_match(conn: &Connection, terms: &[String]) -> Option<String> {
    if terms.is_empty() || !fuzzy_enabled(conn) {
        return None;
    }
    let vocab: HashMap<String, i64> = conn
        .prepare(VOCAB_SQL)
        .ok()?
        .query_map([], |row| {
            Ok((row.get::<_, String>(0)?, row.get::<_, i64>(1)?))
        })
        .ok()?
        .collect::<Result<_, _>>()
        .ok()?;

    // Split the way the index tokenizer does: a query token like `sindicato?`
    // or `¿qué` is one or more vocabulary words with punctuation around them,
    // and looked up whole it would read as a typo of the word it contains.
    let own: Vec<String> = terms
        .iter()
        .flat_map(|term| {
            normalize_term(term)
                .split(|c: char| !c.is_alphanumeric())
                .filter(|word| !word.is_empty())
                .map(str::to_string)
                .collect::<Vec<_>>()
        })
        .collect();
    let mut variants: Vec<String> = Vec::new();
    for term in &own {
        for variant in pick_variants(term, &vocab) {
            if !own.contains(&variant) && !variants.contains(&variant) {
                variants.push(variant);
            }
        }
    }
    if variants.is_empty() {
        return None;
    }
    Some(
        variants
            .iter()
            .map(|variant| format!("\"{variant}\""))
            .collect::<Vec<_>>()
            .join(" OR "),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    fn vocab(entries: &[(&str, i64)]) -> HashMap<String, i64> {
        entries.iter().map(|(t, d)| (t.to_string(), *d)).collect()
    }

    #[test]
    fn folds_case_and_accents() {
        assert_eq!(normalize_term("ZÁRATE"), "zarate");
        assert_eq!(normalize_term("Peñón"), "penon");
    }

    #[test]
    fn counts_a_transposition_as_one_edit() {
        assert_eq!(edit_distance("trabajadores", "trabaajdores", 2), 1);
        assert_eq!(edit_distance("sindicato", "gobernador", 2), 3);
    }

    #[test]
    fn allows_edits_by_length_and_never_on_digits() {
        assert_eq!(max_edits_for("otiz"), 0);
        assert_eq!(max_edits_for("ortiz"), 1);
        assert_eq!(max_edits_for("sindicato"), 2);
        assert_eq!(max_edits_for("fs12345"), 0);
    }

    // The same cases, with the same real-corpus counts, as fuzzy.test.ts.
    #[test]
    fn picks_the_same_variants_as_the_typescript_side() {
        let words = vocab(&[
            ("sindicato", 269),
            ("sindigato", 3),
            ("sinicato", 2),
            ("sindicatos", 119),
            ("trabajadores", 284),
            ("trabajdores", 1),
            ("obrero", 240),
            ("obreros", 266),
            ("obreo", 1),
        ]);
        assert_eq!(
            pick_variants("sindicato", &words),
            ["sindigato", "sinicato"]
        );
        assert_eq!(pick_variants("obrero", &words), ["obreo"]);
        assert_eq!(pick_variants("trabajdores", &words), ["trabajadores"]);
        assert_eq!(pick_variants("sindicto", &words), ["sindicato"]);

        let names = vocab(&[
            ("crocitto", 9),
            ("crocito", 3),
            ("crosito", 1),
            ("crositto", 1),
        ]);
        assert_eq!(
            pick_variants("Crositto", &names),
            ["crocitto", "crosito", "crocito"]
        );
        assert_eq!(pick_variants("Crocito", &names), ["crocitto", "crosito"]);
        assert_eq!(pick_variants("Crosito", &names), ["crocito", "crositto"]);
        assert_eq!(pick_variants("crocitoo", &names), ["crocitto"]);
    }

    /// Two implementations of one rule drift silently unless something reads
    /// both. This reads the TypeScript constants and thresholds.
    #[test]
    fn the_rules_match_fuzzy_ts() {
        let ts = include_str!("../../../../../packages/store/src/fuzzy.ts");
        let constant = |name: &str| -> i64 {
            let line = ts
                .lines()
                .find(|line| line.starts_with(&format!("const {name} = ")))
                .unwrap_or_else(|| panic!("fuzzy.ts no longer declares {name}"));
            line.rsplit(' ').next().unwrap().parse().unwrap()
        };
        assert_eq!(constant("WEIGHT_RATIO"), WEIGHT_RATIO);
        assert_eq!(constant("MAX_MISREADINGS") as usize, MAX_MISREADINGS);
        assert_eq!(constant("RARE_DOCS"), RARE_DOCS);
        assert!(ts.contains("if (term.length < 5) return 0"));
        assert!(ts.contains("if (term.length < 8) return 1"));
        // And both read the vocabulary through the same filter.
        let repo = include_str!("../../../../../packages/store/src/repos/fts.repo.ts");
        let squash = |sql: &str| sql.split_whitespace().collect::<Vec<_>>().join(" ");
        assert!(
            squash(repo).contains(&squash(VOCAB_SQL)),
            "fts.repo.ts no longer reads the vocabulary as VOCAB_SQL does"
        );
    }
}
